### ISSUE TITLE
Polyline marker rendering issue

### ISSUE THREAD

**Issue by sheikh-iffi** [tangrams/tangram-es#1463]
Hi guys

I am facing two problems regarding polyline marker. 

First is that polyline marker doesn't take variable width. I am specifying stops in the styling string but its not working. 
Second issue is that if I give an outline to polyline marker, it disappears at some levels.
In the below images, my polyline style string was:

`std::string polylineStyle = "{ style: lines, interactive: true, color: red, width: [[10,10px], [20,60px]], order: 5000, outline:color: blue, outline:width: 2px }";`

You can see that there is no change of width in polyline and outline has disappeared.

![marker](https://cloud.githubusercontent.com/assets/12157369/25213914/96c289b6-25ad-11e7-99c0-576468ac7487.png)
![marker2](https://cloud.githubusercontent.com/assets/12157369/25213915/99508674-25ad-11e7-8921-55888c973017.png)
![marker3](https://cloud.githubusercontent.com/assets/12157369/25213918/9b868e02-25ad-11e7-93db-0fdf18416cf4.png)

**Comment by nvkelso** [comment]
You might try modifying the nesting of the outline properties?

std::string polylineStyle = "{ style: lines, interactive: true, color: red,
width: [[10,10px], [20,60px]], order: 5000, outline: { color: blue, width:
2px } }";

**Comment by sheikh-iffi** [comment]
Same behavior as before. Outline vanishes after some zoom levels.
![marker4](https://cloud.githubusercontent.com/assets/12157369/25214779/c05bacd0-25b2-11e7-9671-479c68c253dd.png)
![marker5](https://cloud.githubusercontent.com/assets/12157369/25214780/c2122dec-25b2-11e7-82bc-023be0496d61.png)

**Comment by tallytalwar** [comment]
I am able to repro this issue. Need to check more as to what is causing this to happen.

**Comment by tallytalwar** [comment]
Ahh, so marker interface is known to have this issue, its because of precision errors when marker geometry spans a large world area. Its recommended not to use marker interface when the geometry being drawn spans a large area and use the ClientDataSource (`MapData`) interface instead which tiles the data appropriately and hence does not exhibit the same precision limit issue.

You can refer to the documentation for the same here: https://mapzen.com/documentation/tangram/android-sdk/0.5.1/com/mapzen/tangram/MapData.html.

I will investigate on why stops are not working for markers though.

**Comment by tallytalwar** [comment]
I am not able to reproduce the issue for stops not working for line widths for marker styling.

As an example I had the following styling string: `{ style: lines, interactive: true, color: red, width: [[0,1px], [20,100px]], order: 5000 }` which got me the following results at zoom 16 and 20.5 respectively.

![screen shot 2017-04-20 at 1 04 59 pm](https://cloud.githubusercontent.com/assets/360641/25243233/573549ec-25ca-11e7-9ef9-280ddc5372f1.png)

![screen shot 2017-04-20 at 1 05 09 pm](https://cloud.githubusercontent.com/assets/360641/25243240/59f908bc-25ca-11e7-9e46-b730a12f6a17.png)

**Comment by tallytalwar** [comment]
Kindly update if this doesn't work for you. Else we can close this issue in favor of #994 to discuss/address the precision issues with marker drawing.

**Comment by sheikh-iffi** [comment]
I used your styling string but the problem is still here
![marker2](https://cloud.githubusercontent.com/assets/12157369/25263942/b7404776-267c-11e7-96ef-63e53275b664.png)
![marker1](https://cloud.githubusercontent.com/assets/12157369/25263943/b7f10df4-267c-11e7-9f53-d24dfec2188d.png)

But when I draw small polyline markers then i can see that width is changing
![marker3](https://cloud.githubusercontent.com/assets/12157369/25264003/29d7dee8-267d-11e7-8fc3-74f5ad5adb6d.png)
![marker4](https://cloud.githubusercontent.com/assets/12157369/25264008/2bea495a-267d-11e7-9bf4-7913f3af79a5.png)

**Comment by sheikh-iffi** [comment]
@tallytalwar as per your suggestion I used ClientDataSource (MapData) to draw my polyline but there is a problem with it as well. Polyline breaks or vanishes at some point by zooming in. 

Line is there at this level:
![screenshot_20170421-161831](https://cloud.githubusercontent.com/assets/12157369/25275479/ee5af206-26ae-11e7-8f82-c0ea4c0cac56.png)

Line vanishes on zooming in.
![screenshot_20170421-161659](https://cloud.githubusercontent.com/assets/12157369/25275484/f7d7af54-26ae-11e7-937a-2dcef7fa66b4.png)

Similarly
![screenshot_20170421-161812](https://cloud.githubusercontent.com/assets/12157369/25275516/260ea8b4-26af-11e7-9624-dbeb7bef087b.png)
![screenshot_20170421-161818](https://cloud.githubusercontent.com/assets/12157369/25275524/2d524c84-26af-11e7-864e-6980b9f1d45c.png)

I think its related to this issue as well #1444

**Comment by sheikh-iffi** [comment]
@tallytalwar I was debugging to see where is the problem when polyline marker spans a large area and found out that the width equals to zero due to rounding. When width is zero, mesh is not recreated. I tried to remove rounding off and changed data type from int to float but that created new problems.

About  ClientDataSource (MapData), I am still trying to figure out why some lines vanishes. Haven't find anything on it so far.

**Comment by tallytalwar** [comment]
@sheikh-iffi Thanks for debugging this. With markers we are aware of the precision loss, hence I recommended to use ClientDataSource. For markers we will have to rethink our implementation approach to solve the underlying precision issues which creep in when dealing with large geographical extent of a marker geometry.

I still have to debug the issue you showed above with ClientDataSource(s) having discontinuous geometry at different zoom levels. Will spend some time in the following week to debug this issue further, and update the issue my finding and try to have a fix for it.

**Comment by tallytalwar** [comment]
@sheikh-iffi this should be fixed by #1517. We just merged this to master, could you give it a try? Thanks

**Comment by sheikh-iffi** [comment]
@tallytalwar clientDataSource is working fine now so I'll use it for the time being. 
I was using markers as it was easy to update its style without removing and re-adding the feature. Is there any way we could do this with clientDataSource. e.g. I want to change color of polyline from green to red without removing it and then adding it back.

**Comment by tallytalwar** [comment]
Hey @sheikh-iffi 

SInce `clientDataSource` relies on the styling defined in the style sheet, you can apply scene updates using [MapController::queueSceneUpdate](https://mapzen.com/documentation/tangram/android-sdk/0.7.0/com/mapzen/tangram/MapController.html#queueSceneUpdate-java.util.List-) and [MapController::applySceneUpdates](https://mapzen.com/documentation/tangram/android-sdk/0.7.0/com/mapzen/tangram/MapController.html#applySceneUpdates--). This should take care of updating `MapData` styling without removing or readding these MapData sources.

Should I go ahead and close this issue? We are tracking marker precision loss issue in #994

**Comment by sheikh-iffi** [comment]
Well I have used  MapController::queueSceneUpdate and MapController::applySceneUpdates but I don't think it would work in my case. e.g. I have a polyline with multiple segments. Now I want to change the color or width of one segment based on interactivity. How would I do that? Because if I use above methods it will change all the segments of the polyline.

**Comment by Guillaume227** [comment]
A potential problem linked to using applySceneUpdates in your use case is the latency and flicker seen as the whole scene gets redrawn: if you make frequent updates to your segments it might become quite annoying from a UX perspective.

**Comment by tallytalwar** [comment]
@Guillaume227 Agreed.

Other approach with the current interface is that you can add properties to your `MapData` which you can filter and style accordingly in your scene file. We do that currently in our bubble-wrap scene file (https://github.com/tangrams/bubble-wrap/blob/gh-pages/bubble-wrap-style.yaml#L1195-L1209). But if you need to update a property of a `MapData` the only way is to delete and readd this marker with the updated property. We do plan to address these limitations though.

**Comment by code-mayank** [comment]
@sheikh-iffi @tallytalwar 
I'm facing same issue with polyline marker, disappearance at some zoom level, as pointed out the problem was the 'width' value encoded in vertex attributes runs out of precision and gets rounded to zero: https://github.com/tangrams/tangram-es/blob/master/core/src/style/polylineStyle.cpp#L154
So, instead of glm::round, i used glm::ceil, and it's working well for me.

**Comment by sheikh-iffi** [comment]
@code-mayank 
I don't think it would still work well because you are forcing the widths to be 1 when they are less than 0.5. It might be okay if the only problem is that of width only. For example in the below images you can see that width doesn't change on zooming in but it does change when I used ceil. But another problem is that you can see that they are not aligned.(With the same coordinates, it is drawn perfect while using MapData but MapData causes other problems for my application.)
Before: 
![screenshot from 2017-07-27 10-49-21](https://user-images.githubusercontent.com/12157369/28656153-b53401be-72b9-11e7-808f-b5492cbccfa3.png)

 After:
![screenshot from 2017-07-27 10-52-17](https://user-images.githubusercontent.com/12157369/28656170-ca51c6a8-72b9-11e7-8a34-daf3e3361a3a.png)

I tried changing [position](https://github.com/tangrams/tangram-es/blob/master/core/src/style/polylineStyle.cpp#L38) to use ceil as well and it did this:
![screenshot from 2017-07-27 11-07-49](https://user-images.githubusercontent.com/12157369/28656531-ed2ed18c-72bb-11e7-8dde-6b10f27f4e51.png)

Which is ok but this is not a good approach as it doesn't solve all problems. For example:

Before: 
![before](https://user-images.githubusercontent.com/12157369/28656652-9391e9a6-72bc-11e7-88f1-957acf2eb88e.png)

After: 
![after](https://user-images.githubusercontent.com/12157369/28656659-9f5750c8-72bc-11e7-8660-316b6e27ef5e.png)

**Note: These problem arises when you have long polyline**

**Comment by sheikh-iffi** [comment]
@tallytalwar one think I noticed is that if I draw point markers with same coordinates they are drawn perfectly whereas they are displaced when line is drawn. Is there any problem with implementaion [here](https://github.com/tangrams/tangram-es/blob/master/core/src/marker/markerManager.cpp#L213) when projecting?
![points](https://user-images.githubusercontent.com/12157369/28860642-d526d6a6-7776-11e7-96b9-035e15f91776.png)
![line](https://user-images.githubusercontent.com/12157369/28860643-d7b49f7a-7776-11e7-9af5-519dad84a4bd.png)

**Comment by tallytalwar** [comment]
@sheikh-iffi Not exactly. The issue here is when drawing line geometry markers, we run out of floating point precision when the line length crosses a certain mercator meters. Its how we encode out geometry data!

We have this in works for our q3 tasks and will be providing a different marker interface to avoid such issues. Stay tuned.

**Comment by m-sasha** [comment]
Any ETA for this bug? It's a blocker for us.

**Comment by matteblair** [comment]
@m-sasha The best solution for this issue would be to use the `MapData` interface for polylines like this instead of `Marker` (check out the [Android API docs](https://mapzen.com/documentation/tangram/android-sdk/0.9.0/) for `MapController$addDataLayer` and `MapData`).

**Comment by m-sasha** [comment]
@matteblair If I understand the API correctly, that workaround doesn't work for us. We need to allow dynamic styling of the polyline, including even animating it.

**Comment by matteblair** [comment]
Polylines can have animated styling using either `Marker` or `MapData`, this is usually implemented by writing some custom shader logic that uses time as an input.

Polylines cannot have their coordinates animated using either `Marker` or `MapData`. If this is a requirement for your application then you'll need your own solution for rendering these polylines.
